### PR TITLE
SALTO-845 - Include fields from the metadata API in custom objects

### PR DESCRIPTION
**Pull request by netama**
We're seeing that not all fields are returned in the `DescribeSObject` response. Some of the reasons are:
1. Only fields that the querying user has access to are returned - so the response can vary even for standard objects and even if all users are admins.
2. Some fields, like `Asset.DigitalAssetStatus`, are just never returned even if they are readable by the user.

When testing on a sample dev workspace with an admin account, this added ~90 new fields (which also resulted in more success in generating references for existing data).

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 1]
```diff
@@ -253,6 +253,23 @@ const mergeCustomObjectWithInstance = (
   customObject: ObjectType, fieldNameToFieldAnnotations: Record<string, Values>,
   instance: InstanceElement, annotationTypesFromInstance: TypeMap
 ): void => {
+  // add fields that exist in the metadata API but not in the SObject
+  const instanceFields = makeArray(instance.value.fields)
+  const objectName = instance.value[INSTANCE_FULL_NAME_FIELD]
+  instanceFields
+    .forEach((field: Values) => {
+      const fieldFullName = naclCase(field[INSTANCE_FULL_NAME_FIELD])
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:261]: we're not always using `naclCase` (we use it [here](https://github.com/salto-io/salto/blob/d4fea9a76a7abbba01786b295f2dfb77aa9581b7/packages/salesforce-adapter/src/filters/custom_objects.ts#L355) but not [here](https://github.com/salto-io/salto/blob/d4fea9a76a7abbba01786b295f2dfb77aa9581b7/packages/salesforce-adapter/src/filters/custom_objects.ts#L404)) - is it needed in this case / in all cases?

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 2]
```diff
@@ -253,6 +253,23 @@ const mergeCustomObjectWithInstance = (
...
+      if (customObject.fields[fieldFullName] === undefined) {
+        log.debug(`Adding field ${fieldFullName} under ${objectName} from metadata API that was missing in SObject`)
+        customObject.fields[fieldFullName] = new Field(
+          customObject,
+          fieldFullName,
+          getFieldType(getFieldName(field)),
+          transformFieldAnnotations(field, objectName),
+        )
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:269]: Some fields have very little information - probably worth discussing which annotations we want to add and how to infer them (or decide on the minimal data needed in order to include the field). I added some examples on the ticket.

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 3]
```diff
@@ -253,6 +253,23 @@ const mergeCustomObjectWithInstance = (
   customObject: ObjectType, fieldNameToFieldAnnotations: Record<string, Values>,
   instance: InstanceElement, annotationTypesFromInstance: TypeMap
 ): void => {
+  // add fields that exist in the metadata API but not in the SObject
+  const instanceFields = makeArray(instance.value.fields)
+  const objectName = instance.value[INSTANCE_FULL_NAME_FIELD]
+  instanceFields
+    .forEach((field: Values) => {
+      const fieldFullName = naclCase(field[INSTANCE_FULL_NAME_FIELD])
+      if (customObject.fields[fieldFullName] === undefined) {
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:262]: working on making this condition more granular - will update when it's ready for review

File: packages/salesforce-adapter/src/filters/custom_objects.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:261]: pretty sure we shouldn't be using `naclCase` for anything that is related to an ID - it isn't 1:1 mapping thus it can cause collisions. took us a while to figure that out, that first example you found is probably a mistake and should be removed

File: packages/salesforce-adapter/src/filters/custom_objects.ts (hunk 2 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:269]: at the very least we should have `apiName` (otherwise references won't work).
probably worth checking if the fields we don't get from SOAP are useable in SOQL, I assume they won't be (because we don't have permission to see them), so maybe we need to add a "readable" annotation (like we have "creatable" and "updatable") that we'll set to `false` for fields that we did not get from the SOAP API

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 4]
```diff
@@ -249,10 +250,49 @@ const transformObjectAnnotations = (customObject: ObjectType, annotationTypesFro
...
 }
 
+const createFieldFromAnnotations = (
+  customObject: ObjectType,
+  field: Values,
+  instance: InstanceElement,
+): Field => {
+  // TODO remove - tmp for demonstration
+  // if (!field[INSTANCE_TYPE_FIELD]) {
+  //   field[INSTANCE_TYPE_FIELD] = 'Text'
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:260]: Most of the standard fields return from the metadata API without a type, so they also don't get an api name and can't be used (moving the discussion to the ticket).

File: packages/salesforce-adapter/src/constants.ts [hunk 5]
```diff
@@ -48,6 +48,8 @@ export enum FIELD_TYPE_NAMES {
   LOOKUP = 'Lookup',
   MASTER_DETAIL = 'MasterDetail',
   ROLLUP_SUMMARY = 'Summary',
+  // internal-only placeholder for fields whose type is unknown
+  UNKNOWN = 'Unknown',
```
Comment (netama) [inline packages/salesforce-adapter/src/constants.ts:52]: @ori-moisis based on our conversation I didn't add defenses against deploy attempts and will let them fail on salesforce errors (`soapenv:Client: 'Unknown' is not a valid value for the enum 'FieldType'`) - lmk if that changes.

File: packages/salesforce-adapter/src/constants.ts [hunk 6]
```diff
@@ -159,6 +161,8 @@ export const FIELD_ANNOTATIONS = {
   FORMULA_TREAT_BLANKS_AS: 'formulaTreatBlanksAs',
   CREATABLE: 'createable',
   UPDATEABLE: 'updateable',
+  // indicates whether a field is queryable by SOQL (default true)
+  QUERYABLE: 'queryable',
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/constants.ts:165]: in `listMetadataTypes` there are `apiCreatable`, `apiUpdatable`, `apiReadable` and `apiDeletable`
I think calling this `readable` would make more sense but not super critical

File: packages/salesforce-adapter/src/constants.ts (hunk 5 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/constants.ts:52]: I think we might need to filter out these fields when the user tries to create a new CustomObject for example because while I generally think this is a good enough error if the user explicitly tries to change the field, I don't think failing like this makes sense if the user tried to create a new custom object (most likely this can happen if the user tries to move a custom object from one environment to another).
The [change validator mechanism](https://github.com/salto-io/salto/blob/master/packages/salesforce-adapter/src/change_validators/picklist_standard_field.ts) should be a good fit for this kind of filtering (and it will also give us the opportunity to give a more readable error message)

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 7]
```diff
@@ -249,10 +249,48 @@ const transformObjectAnnotations = (customObject: ObjectType, annotationTypesFro
...
+    .filter((field: Values) => (customObject.fields[field[INSTANCE_FULL_NAME_FIELD]] === undefined))
+    .map((field: Values): Field => createFieldFromAnnotations(customObject, field, instance))
+    .filter((fieldElem: Field) => (apiName(fieldElem) !== undefined))
+
+  if (fieldsToAdd.length > 0) {
+    log.debug(`Extending SObject ${apiName(customObject)} with fields from metadata API: ${
+      fieldsToAdd.map(f => f.name)
+    }`)
+    Object.assign(customObject.fields, ...fieldsToAdd.map((f: Field) => ({ [f.name]: f })))
+  }
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:292]: `fieldNameToFieldAnnotations` already holds a mapping of the instance fields by name and the code in `transformFieldAnnotations` already doesn't do anything for fields that are not defined in the instance (see [this part in the code](https://github.com/salto-io/salto/blob/master/packages/salesforce-adapter/src/filters/custom_objects.ts#L208-L216)).

I think it would make more sense to change the loop here so that instead of running `forEach` on the current fields, it will run `forEach` on the fields from the instance.
something along the lines of
```typescript
  const fieldByApiName = _.mapKeys(customObject.fields, field => apiName(field, true))
  Object.entries(fieldNameToFieldAnnotations).forEach(([name, values]) => {
    const metadataAnnotations = transformFieldAnnotations(
      values,
      instance.value[INSTANCE_FULL_NAME_FIELD], // btw - I think we can get this as "parentName" parameter instead of getting the whole instance
    )
    const field = fieldByApiName[name]
    if (field !== undefined) { // maybe swap the order of conditions?
      Object.assign(field.annotations, metadataAnnotations)
    } else {
      customObject.fields[name] = new Field(
        customObject,
        name,
        getFieldType(values[INSTANCE_TYPE_FIELD]),
        metadataAnnotations,
      )
    }
  })
```

Of course this would require some changes in `transformFieldAnnotations` and probably also `getFieldType` but I think those changes would make the code better overall because currently we have one place in the code (`transformFieldAnnotations`) that says we ignore fields if they don't have a type and another place in the code (this new code) that handles fields that don't have a type

btw, the additional part of the loop where the current code runs
```typescript
    if (field.annotations[FIELD_ANNOTATIONS.VALUE_SET]
      && field.annotations[VALUE_SET_FIELDS.VALUE_SET_NAME]) {
      delete field.annotations[FIELD_ANNOTATIONS.VALUE_SET]
    }
```
can stay in a separate loop since it doesn't really have anything to do with merging the definitions IMO

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 8]
```diff
@@ -249,10 +249,48 @@ const transformObjectAnnotations = (customObject: ObjectType, annotationTypesFro
...
+  if (!field[INSTANCE_TYPE_FIELD]) {
+    field[INSTANCE_TYPE_FIELD] = FIELD_TYPE_NAMES.UNKNOWN
+  }
+  const annotations = transformFieldAnnotations(
+    field,
+    instance.value[INSTANCE_FULL_NAME_FIELD],
+  )
+  annotations[FIELD_ANNOTATIONS.QUERYABLE] = false
+  annotations[FIELD_ANNOTATIONS.UPDATEABLE] = false
+  annotations[FIELD_ANNOTATIONS.CREATABLE] = false
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:266]: With the current code I think the `READABLE` annotation behaves differently from `UPDATEABLE` and `CREATEABLE`.
the latter always have a value but `READABLE` only has a value it if is false.
I don't really have a strong preference for one behavior over the other, but I do think it should be the same behavior for all of these very similar annotations.
I think that either `READABLE` should be added to `getSObjectFieldElement`, or the behavior of `UPDATEABLE` and `CREATABLE` should be changed to match `READABLE`

File: packages/salesforce-adapter/src/constants.ts [hunk 9]
```diff
@@ -48,6 +48,8 @@ export enum FIELD_TYPE_NAMES {
   LOOKUP = 'Lookup',
   MASTER_DETAIL = 'MasterDetail',
   ROLLUP_SUMMARY = 'Summary',
+  // internal-only placeholder for fields whose type is unknown
```
[the remark this answers is not in the document]
Comment (netama) [reply packages/salesforce-adapter/src/constants.ts:51]: that it's not a real type and should not be sent to the service. happy to change the phrasing if you have suggestions :)

File: packages/salesforce-adapter/src/change_validators/unknown_field.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  isField(changedElement)
+  && changedElement.type.elemID.isEqual(Types.primitiveDataTypes.Unknown.elemID)
+)
+
+const createChangeError = (field: Field): ChangeError =>
+  ({
+    elemID: field.elemID,
+    severity: 'Error',
+    message: `You cannot create or modify a field with unknown type. Field: ${field.name}`,
+    detailedMessage: 'You cannot create or modify a field with unknown type. Check your profile permissions on Salesforce to make sure your have access to the field.',
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/change_validators/unknown_field.ts:32]: I'm always confused by what is supposed to be the `message` and what is the `detailedMessage`
@omrilit - is this the correct usage?

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 11]
```diff
@@ -1103,14 +1112,17 @@ export const getSObjectFieldElement = (
     _.assign(
       annotations,
       _.pick(
-        _.omit(field, Object.keys(annotations)),
+        Object.assign(
+          _.omit(field, Object.keys(annotations)),
+          { [FIELD_ANNOTATIONS.QUERYABLE]: true },
+        ),
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/transformers/transformer.ts:1118]: 1. why is this under this `if`? doesn't seem related to what is supposed to happen here (as the comment says, this is where we take everything from the field definition that has the same name, `QUERYABLE` is a name we made up)
2. this is an extremely indirect way to write
```
annotations[FIELD_ANNOTATIONS.QUERYABLE] = true
```
why?

Comment (netama) [reply packages/salesforce-adapter/src/transformers/transformer.ts:1118]: my understanding from your [previous comment](https://github.com/salto-io/salto/pull/1240/files#r462155154) was that it should follow the pattern of `createable` / `updateable` - this mapping is what sets them, and it's only done when they're listed in the field type annotations. wasn't the goal to align this annotation with them?

Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/transformer.ts:1118]: so I think what I've been missing so far in all of this PR is that `queryable` is already being set on all fields once you added it to the annotation types because it is in the reply from the SOAP API
This means that this code should definitely be deleted because it overrides correct values with potentially incorrect values

Comment (netama) [reply packages/salesforce-adapter/src/transformers/transformer.ts:1118]: I'm not sure what you mean. this is a new annotation, and even if it exists in the API response the `_.omit()` part ensures we don't override it.

Comment (netama) [reply packages/salesforce-adapter/src/transformers/transformer.ts:1118]: I don't follow - it's not returned, and if we don't set it here it doesn't get initialized (I double checked again now). let's discuss today. if we are worried I can change the `assign` order

File: packages/salesforce-adapter/src/constants.ts (hunk 6 above)
Comment (netama) [reply packages/salesforce-adapter/src/constants.ts:166]: discussed IRL - going to use `queryable` for now

File: packages/salesforce-adapter/src/transformers/transformer.ts (hunk 11 above)
Comment (netama) [reply packages/salesforce-adapter/src/transformers/transformer.ts:1118]: discussed IRL - it appears on the object but not the fields so there's no risk of override - for now for simplicity we will always set it regardless of whether it's configured on the field annotations or not (so moving outside the `if`)

File: packages/salesforce-adapter/src/filters/custom_objects.ts (hunk 7 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:292]: I looked at it again and I'm not sure I agree - I think the distinction between fields that exist in SOAP and those that are initialized from the metadata API is important, and shouldn't be hidden under `transformFieldAnnotations`. we can add something clearer at the top-level to avoid attempting to call it again, but I do think it should stay visible.

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 12]
```diff
@@ -249,20 +249,60 @@ const transformObjectAnnotations = (customObject: ObjectType, annotationTypesFro
     transformObjectAnnotationValues(instance, annotationTypesFromInstance))
 }
 
+const createFieldFromAnnotations = (
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:252]: should this be `createFieldFromMetadataInstance` or something like that?
calling it `createFieldFromAnnotations ` is a bit confusing because the source of information for the field is not annotations, at least as far as I understand

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 13]
```diff
@@ -249,20 +249,60 @@ const transformObjectAnnotations = (customObject: ObjectType, annotationTypesFro
...
+  customObject: ObjectType,
+  field: Values,
+  instanceName: string,
+): Field => {
+  if (!field[INSTANCE_TYPE_FIELD]) {
+    field[INSTANCE_TYPE_FIELD] = FIELD_TYPE_NAMES.UNKNOWN
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:258]: we shouldn't change function inputs, causing side effects tends to be bad for readability.
If this is only here because of the call to `getFieldType(getFieldName(field))` later on, then maybe it would be better to avoid this altogether?
if not then this is even more confusing than I thought

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 14]
```diff
@@ -249,20 +249,60 @@ const transformObjectAnnotations = (customObject: ObjectType, annotationTypesFro
...
+      // doesn't exist in SOAP - initialize from metadata API
+      const field = createFieldFromAnnotations(customObject, values, instanceName)
+      if (apiName(field) !== undefined) {
+        log.debug(`Extending SObject ${apiName(customObject)} with field ${fieldName} from metadata API`)
+        customObject.fields[fieldName] = field
+      }
     }
   })
+
+  _(customObject.fields).filter(field => (
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:301]: nit: prefer built in functionality over using external libraries
```suggestion
  Object.values(customObject.fields).filter(field => (
```

File: packages/salesforce-adapter/src/filters/custom_objects.ts (hunk 13 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:258]: It's here because this is needed in a few places downstream, the first one is in [`transformFieldAnnotations`](https://github.com/salto-io/salto/blob/fa58cbac3da09a16b9e0557938ce135de413aef9/packages/salesforce-adapter/src/filters/custom_objects.ts#L209) (and again a few lines later). I assumed since what we do in many of the filters is modify the elements in-place (and they're also passed in as input parameters), it was ok to do here as well. What's the rule of thumb for when it's not ok? And if we want to avoid it, will cloning be a reasonable solution? I would rather not keep modifying all the underlying functions in this filter (it may be worth rewriting but then I think we should do it as its own task), and I think passing it as an additional parameter will make the code unnecessarily complex.

Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/custom_objects.ts:258]: I think generally we should always try to avoid side effects (changing input) in functions.
If we have to change input, it should be the only thing the function does, I think a reasonable way to tell if it is ok is that the function that changes its input should have no return value and ideally the name of the function should hint heavily towards what it does.
I think in this case I would have no reason to suspect that calling this function that by its name and return value seems to create a field should also be the part of the code that is responsible for filling in a type in metadata instances if it is missing.

personally I would have preferred to change the other functions to fit with the new code (they are still quite confusing IMO, stating explicitly that we don't support fields without types when a lot of this new code deals exactly with supporting that) but building a different input would work as well.

just don't use `_.cloneDeep`
